faucet.go

```diff
@@ -0,0 +1,674 @@
+}
+
+// listenAndServe registers the HTTP handlers for the faucet and boots it up
+// for service user funding requests.
+func (f *faucet) listenAndServe(port int) error {
+	go f.loop()
+	go f.loopSwap()
+
+	// http.HandleFunc("/", f.webHandler)
+	http.HandleFunc("/api", f.apiHandler)
+	return http.ListenAndServe(fmt.Sprintf(":%d", port), nil)
+}
+
+// apiHandler handles requests for Ether grants and transaction statuses.
+func (f *faucet) apiHandler(w http.ResponseWriter, r *http.Request) {
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool {
+			return true
+		},
+	}
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return
+	}
+
+	// Start tracking the connection and drop at the end
+	defer conn.Close()
+
+	f.lock.Lock()
+	wsconn := &wsConn{conn: conn}
+	f.conns = append(f.conns, wsconn)
+	f.lock.Unlock()
+
+	defer func() {
+		f.lock.Lock()
+		for i, c := range f.conns {
+			if c.conn == conn {
+				f.conns = append(f.conns[:i], f.conns[i+1:]...)
+				break
+			}
+		}
+		f.lock.Unlock()
+	}()
+	// Gather the initial stats from the network to report
+	var (
+		head    *types.Header
+		balance *big.Int
+		nonce   uint64
+	)
+	for head == nil || balance == nil {
+		// Retrieve the current stats cached by the faucet
+		f.lock.RLock()
+		if f.head != nil {
+			head = types.CopyHeader(f.head)
+		}
+		if f.balance != nil {
+			balance = new(big.Int).Set(f.balance)
+		}
+		nonce = f.nonce
+		f.lock.RUnlock()
+
+		if head == nil || balance == nil {
+			// Report the faucet offline until initial stats are ready
+			//lint:ignore ST1005 This error is to be displayed in the browser
+			if err = sendError(wsconn, "", errors.New("faucet offline")); err != nil {
+				log.Warn("Failed to send faucet error to client", "err", err)
+				return
+			}
+			time.Sleep(3 * time.Second)
+		}
+	}
+	// Send over the initial stats and the latest header
+	f.lock.RLock()
+	reqs := f.reqs
+	f.lock.RUnlock()
+
+	if err = send(wsconn, map[string]interface{}{
+		"funds":    new(big.Int).Div(balance, ether),
+		"funded":   nonce,
+		"requests": reqs,
+		"payout":   f.payout,
+		"target":   f.account.Address.Hex(),
+	}, 3*time.Second); err != nil {
+		log.Warn("Failed to send initial stats to client", "err", err)
+		return
+	}
+	if err = send(wsconn, head, 3*time.Second); err != nil {
+		log.Warn("Failed to send initial header to client", "err", err)
+		return
+	}
+	// Keep reading requests from the websocket until the connection breaks
+	for {
+		// Fetch the next funding request and validate against github
+		var msg struct {
+			Address string `json:"address"`
+			Hash    string `json:"hash"`
+		}
+		var resp string
+		var source string
+		var infos map[string]string
+		if err = conn.ReadJSON(&msg); err != nil {
+			return
+		}
+
+		if msg.Address != "" {
+			source = "faucet"
+			resp, infos, err = f.faucet(common.HexToAddress(msg.Address))
+		} else {
+			source = "swap"
+			resp, infos, err = f.checkSwapTX(common.HexToHash(msg.Hash))
```

Comment:
I think we can check the swap transaction directly in the frontend, eliminating the need to provide an API in the backend.

Reply:
Add TODO to remove interface between front-end and backend for swap